src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRazorPageFactoryProvider.cs

```diff
@@ -74,15 +74,15 @@ public RazorPageFactoryResult CreateFactory(string relativePath)
         /// <summary>
         /// Creates a factory for <see cref="IRazorPage"/>.
         /// </summary>
-        /// <param name="compiledType">The <see cref="Type"/> to produce an instance of <see cref="IRazorPage"/>
+        /// <param name="pageCreate">The <see cref="Func{Object}"/> to produce an instance of <see cref="IRazorPage"/>
         /// from.</param>
         /// <param name="relativePath">The application relative path of the page.</param>
-        /// <returns>A factory for <paramref name="compiledType"/>.</returns>
-        protected virtual Func<IRazorPage> GetPageFactory(Type compiledType, string relativePath)
+        /// <returns>A factory for <see cref="IRazorPage"/>.</returns>
+        protected virtual Func<IRazorPage> GetPageFactory(Func<object> pageCreate, string relativePath)
         {
             return () =>
             {
-                var page = (IRazorPage)Activator.CreateInstance(compiledType);
+                var page = (IRazorPage)pageCreate();
                 page.Path = relativePath;
                 return page;
             };
```

Comment:
/cc @pranavkm - should we simplify this further now that the compilation result has a `Func<object>`